src/main/java/smartthings/konsumer/filterchain/MessageFilterChain.java

```diff
@@ -10,37 +10,37 @@
 
 import java.util.*;
 
-public class MessageFilterChain {
+public class MessageFilterChain<K, V, R> {
 	private final static Logger log = LoggerFactory.getLogger(MessageFilterChain.class);
 
-	private final MessageProcessor processor;
-	private final List<MessageFilter> filters;
+	private final MessageProcessor<K, V, R> processor;
+	private final List<MessageFilter<K, V, R>> filters;
 
-	public MessageFilterChain(final MessageProcessor processor, final MessageFilter... messageFilters) {
+	public MessageFilterChain(final MessageProcessor<K, V, R> processor, final List<MessageFilter<K, V, R>> messageFilters) {
 		this.processor = processor;
-		this.filters = Collections.unmodifiableList(new ArrayList<MessageFilter>(Arrays.asList(messageFilters)));
+		this.filters = Collections.unmodifiableList(new ArrayList<>(messageFilters)); //make defensive copy
 	}
 
-	public void handle(MessageAndMetadata<byte[], byte[]> originalMessageAndMetadata,
+	public R handle(MessageAndMetadata<K, V> originalMessageAndMetadata,
```

Comment:
I was in doubt about adding this, but it is helpful in some situations. 

What I needed it for, is to be able to calculate error ratios for message processing. In many situations a lot of messages will be thrown away quickly and therefore not be processed at all. The baseline number for calculating the error ratios includes all messages, so even a high percentage of processed events failing will not make a big impact on the error ratios. This allows us to use a return value to signal to the filters if a message was actually processed and should be included in the ratio calculations, or was simply thrown away.